File: src/main/java/com/hohltier/decompiler/menus/file/FileExit.java
package com.hohltier.decompiler.menus.file;

import com.formdev.flatlaf.extras.components.FlatMenuItem;
import com.hohltier.decompiler.controllers.WindowController;
import com.hohltier.decompiler.services.LanguageService;
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;

public class FileExit extends FlatMenuItem implements ActionListener {

    public FileExit() {
        addActionListener(this);
        setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_Q, Toolkit.getDefaultToolkit().getMenuShortcutKeyMaskEx()));
        setMnemonic(KeyEvent.VK_Q);
        setText(LanguageService.getTranslation("file.exit"));

        if (Desktop.isDesktopSupported() && Desktop.getDesktop().isSupported(Desktop.Action.APP_QUIT_HANDLER)) {
            setVisible(false);
            Desktop.getDesktop().setQuitHandler((e, r) -> actionPerformed(null));
        }
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        WindowController.getINSTANCE().dispose();
    }

}

File: src/main/kotlin/moe/sota/decompiler/menus/file/FileExit.kt
package moe.sota.decompiler.menus.file

import com.formdev.flatlaf.extras.components.FlatMenuItem
import moe.sota.decompiler.controllers.WindowController
import moe.sota.decompiler.services.LanguageService
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import java.awt.Desktop
import java.awt.Toolkit
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.KeyEvent
import javax.swing.KeyStroke

class FileExit : FlatMenuItem(), ActionListener, KoinComponent {

    private val windowController: WindowController by inject()

    init {
        accelerator = KeyStroke.getKeyStroke(KeyEvent.VK_Q, Toolkit.getDefaultToolkit().menuShortcutKeyMaskEx)
        isVisible = !(Desktop.isDesktopSupported() && Desktop.getDesktop().isSupported(Desktop.Action.APP_QUIT_HANDLER))
        mnemonic = KeyEvent.VK_Q
        text = LanguageService.getTranslation("file.exit")
        addActionListener(this)

        if (!isVisible)
            Desktop.getDesktop().setQuitHandler { _, _ -> actionPerformed(null) }
    }

    override fun actionPerformed(e: ActionEvent?) {
        windowController.dispose()
    }

}
